### ISSUE TITLE
Bidirectional hasMany relationships are not loadable/saveable from the many side

### ISSUE THREAD

**Issue by rsutphin** [nolanlawson/ember-pouch#16]
The default behavior of `DS.JSONSerializer` is to not serialize the many side of bidirectional hasMany relationships. See emberjs/data#2494 for links to discussion about this.

The upshot for `ember-pouch` is that the contents of a bidirectional hasMany relationship isn't saved or loaded when working from the many side. E.g., say you have these models:

```javascript
// app/models/person.js
export default DS.Model.extend({
  sandwiches: DS.hasMany('sandwich', { inverse: 'owner' })
});

// app/models/sandwich.js
export default DS.Model.extend({
  owner: DS.belongsTo('person')
});
```

In this case you won't be able to load or save a person's sandwiches from the person. (You will be able to if you remove `owner` from sandwich, though.)

The workaround for this that other similar adapters use ([ember-localstorage-adapter][LS], [ember-indexeddb-adapter][IDB]) is to provide a serializer which forces these associations to be serialized on the hasMany side (as well as the belongsTo side). I'm doing this now in my own application to work around this issue.

[LS]: https://github.com/kurko/ember-localstorage-adapter/blob/c32477626ce0e7c8447e1ee79eb78e64a8357d70/localstorage_adapter.js#L6
[IDB]: https://github.com/kurko/ember-indexeddb-adapter/blob/0b1c73a1134606ccc9dfc1aadb86af5b4dd9fe4d/packages/indexeddb-adapter/lib/indexeddb_serializer.js#L2

Unfortunately this requires copying a bunch of code from JSONSerializer (hence emberjs/data#2494). It also duplicates the information about the relationship, requiring both sides to be saved manually whenever the relationship changes.

I'm not sure if there's a practical better solution. It appears that the ember-data solution is to have the server use different serializations for reads vs. writes, but that doesn't work for a local database. One idea I had was that the adapter could look up the IDs for the hasMany relationship dynamically — since all the data is local this might work, but it also might be complex to get right.

**Comment by fsmanuel** [comment]
I'm not sure when ember-data added `_shouldSerializeHasMany` but in 1.13.x it is available
We can add that line to the serializer and add a section to the readme that ppl should import the serializer.

```js
  _shouldSerializeHasMany: function() { return true; }
```

**Comment by broerse** [comment]
Or use an initializer to import it? Example: https://github.com/firebase/emberfire/blob/master/addon/initializers/emberfire.js

**Comment by fsmanuel** [comment]
I'm more in favor of an explicit import.
We already expose the serializer:
```js
import {Model, Adapter, Serializer} from 'ember-pouch';
```

**Comment by rsutphin** [comment]
I'm in favor of an explicit import also. An initializer is harder for a user of the library to discover & override.

**Comment by broerse** [comment]
If it is well documented for beginners lets do the explicit import.

**Comment by fsmanuel** [comment]
@patuku-roy I'm not 100% on this but you should use a serializer per type that extends your pouch serializer (it's just the RESTSerializer) and declare the `attrs` hash.
Something linke:
```js
export default DS.RESTSerializer.extend(DS.EmbeddedRecordsMixin, {
  attrs: {
    comments: {serialize: true}
  }
});
```
See the docs for more infos about [DS.EmbeddedRecordsMixin](http://emberjs.com/api/data/classes/DS.EmbeddedRecordsMixin.html)

**Comment by patuku-roy** [comment]
@fsmanuel Thanks, I did per type serializer and work like a charm, except I can not use `serialize: true`, I'm using `serialize: "ids"`. The good news, I don't have a problem using ED Beta 18 and I don't have to use application level serializer also.

Another problem coming, every time I saved the model, I can not fetch the record until I refresh the route for couple times, but I suspect this is Ember runloop issue when reload the model rather than ember-pouch issue.

```javascript
export default DS.RESTSerializer.extend(DS.EmbeddedRecordsMixin, {
  attrs: {
    comments: {serialize: "ids"}
  }
});
```

and I save the model like this

```javascript
this.store.createRecord("comment", {
    // ... put the data here ...
}).save().then(function(comment) {
    // save relation to each side
    post.get("comments").get("content").pushObject(comment);
    post.save();
});
```

**Comment by fsmanuel** [comment]
`post.get("comments").pushObject(comment);` should work.

what do you mean by "every time I saved the model, I can not fetch the record until I refresh the route for couple times"?

**Comment by patuku-roy** [comment]
@fsmanuel My bad, I use transitionTo before the record saved, after I'm waiting the record saved, the model can show up without any problem.

This is the working version,

```javascript
// app/routes/post.js
var self = this;

this.store.createRecord("comment", {
    // ... put the data here ...
}).save().then(function(comment) {
    // save relation to each side
    post.get("comments").get("content").pushObject(comment);
    post.save();
    self.set("didSave", true);
});
```

and add new method to use transitionTo when the record saved.

```javascript
// app/routes/post.js

didSave: false,
_transitionAfterSave: function(transitionTo) {
    var self    = this;

    this.addObserver("didSave", function() {
        if(self.get("didSave"))
        {
            self.set("didSave", false);
            self.transitionTo(transitionTo);
        }
    });
},
```

Thanks anyway ...

**Comment by broerse** [comment]
@fsmanuel Do you have time to write a PR for `_shouldSerializeHasMany: function() { return true; }` in the serializer and document it?

**Comment by MSchmidt** [comment]
Is there any progress on this one? After reading through the issue I'm still not sure how to do this the correct way.

**Comment by openhouse** [comment]
I got this to work by adding an application serializer that extends ember-pouch's serializer:

```
//  app/serializers/application.js
import Ember from 'ember';
import { Serializer } from 'ember-pouch';

export default Serializer.extend({
  _shouldSerializeHasMany: function() { return true; }
});
```

Now the ids for the manyToOne relationship are stored on the hasMany side as well.

Here is the private method we are overriding
https://github.com/emberjs/data/blob/v2.2.1/packages/ember-data/lib/serializers/json-serializer.js#L814-L830

Ember: 2.2.0
Ember Data: 2.2.1

**Comment by wooandoo** [comment]
@openhouse thanks for you simple solution!

**Comment by nolanlawson** [comment]
@openhouse could you open a PR? If your method is successful, it should be added to the repo. :)

**Comment by backspace** [comment]
I made a PR with the suggested fix but it doesn’t seem like it works, to me… though maybe I’m missing something?

**Comment by backspace** [comment]
Maybe it _is_ working in my application, but I don’t know why the test in my PR is failing.

This situation where the child record needs to be saved, inserted into the parent collection, and then the parent saved seems unwieldy and confusing for people who are familiar with Ember Data. Is anyone knowledgable enough about Ember Data to know whether there’s a way to cause a parent model to be automatically saved when a child model is? I’ve poked around the code and it seems like maybe it’s not possible, but maybe someone has a better understanding.

**Comment by bitmorse** [comment]
@openhouse thanks a lot! It's a lifesaver!

**Comment by backspace** [comment]
Hey, so I submitted [another PR](https://github.com/nolanlawson/ember-pouch/pull/111) that incorporates the `_shouldSerializeHasMany` fix with a test that passes in all but the Ember 1.13 environment. The problem with my previous PR was that the dummy application wasn’t using the addon’s serialiser.

I’ll have to come back to see why that test is failing; it’s a seemingly-unrelated one.

**Comment by nolanlawson** [comment]
fixed by 3cb223173ca53c3a864ae23db86c79b43c020fa9 and #111
